File: scripts/utils/init_logger.py
#!/usr/bin/env python3
import logging, os, datetime, subprocess, hashlib

def init_logger(name="pipeline", log_dir="results/logs"):
    os.makedirs(log_dir, exist_ok=True)
    timestamp = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
    log_path = os.path.join(log_dir, f"{name}_{timestamp}.log")

    logging.basicConfig(
        filename=log_path,
        level=logging.INFO,
        format="%(asctime)s | %(levelname)s | %(name)s | %(message)s",
        datefmt="%Y-%m-%d %H:%M:%S",
    )

    log = logging.getLogger(name)
    try:
        commit = subprocess.getoutput("git rev-parse --short HEAD")
        env_hash = hashlib.sha256(
            open("environment/conda.lock", "rb").read()
        ).hexdigest()[:10]
        log.info(f"Git commit: {commit}")
        log.info(f"Environment digest: {env_hash}")
    except Exception as e:
        log.warning(f"Metadata logging skipped: {e}")
    return log

File: scripts/utils/init_logger.R
#!/usr/bin/env Rscript

init_logger <- function(name="pipeline", log_dir="results/logs") {
  dir.create(log_dir, showWarnings = FALSE, recursive = TRUE)
  timestamp <- format(Sys.time(), "%Y%m%d_%H%M%S")
  log_file <- file.path(log_dir, paste0(name, "_", timestamp, ".log"))
  log_con <- file(log_file, open="wt")

  info <- function(msg) {
    cat(sprintf("%s INFO %s | %s\n", Sys.time(), name, msg), file=log_con)
    flush(log_con)
  }

  info(sprintf("Git commit: %s",
               tryCatch(system("git rev-parse --short HEAD", intern=TRUE), error=function(e) "N/A")))
  if (file.exists("environment/conda.lock")) {
    digest <- tools::md5sum("environment/conda.lock")
    info(sprintf("Environment digest: %s", substr(digest, 1, 10)))
  }

  info("Logger initialized.")
  structure(list(info=info, file=log_file), class="logger")
}
